File: src/main/java/com/beat/domain/staff/domain/Staff.java
package com.beat.domain.staff.domain;

import com.beat.domain.performance.domain.Performance;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Staff {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	@Column(nullable = false)
	private String staffName;

	@Column(nullable = false)
	private String staffRole;

	@Column(nullable = false)
	private String staffPhoto;

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "performance_id", nullable = false)
	@OnDelete(action = OnDeleteAction.CASCADE)
	private Performance performance;

	@Builder
	private Staff(String staffName, String staffRole, String staffPhoto, Performance performance) {
		this.staffName = staffName;
		this.staffRole = staffRole;
		this.staffPhoto = staffPhoto;
		this.performance = performance;
	}

	public static Staff create(String staffName, String staffRole, String staffPhoto, Performance performance) {
		return Staff.builder()
			.staffName(staffName)
			.staffRole(staffRole)
			.staffPhoto(staffPhoto)
			.performance(performance)
			.build();
	}

	public void update(String staffName, String staffRole, String staffPhoto) {
		this.staffName = staffName;
		this.staffRole = staffRole;
		this.staffPhoto = staffPhoto;
	}
}

File: domain/src/main/kotlin/com/beat/domain/performance/model/Staff.kt
package com.beat.domain.performance.model

class Staff private constructor(
    private val staffId: Id?,
    val staffName: String,
    val staffRole: String,
    val staffPhoto: String,
) {
    val id: Long?
        get() = staffId?.value

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Staff) return false
        return staffId != null && staffId == other.staffId
    }

    override fun hashCode(): Int = staffId?.hashCode() ?: System.identityHashCode(this)

    override fun toString(): String = "Staff(id=$id)"

    fun update(staffName: String, staffRole: String, staffPhoto: String): Staff = Staff(
        staffId = staffId,
        staffName = staffName,
        staffRole = staffRole,
        staffPhoto = staffPhoto,
    )

    @JvmInline
    value class Id private constructor(val value: Long) {
        companion object {
            fun fromNullable(value: Long?): Id? = value?.let(::Id)
        }
    }

    companion object {
        fun create(staffName: String, staffRole: String, staffPhoto: String): Staff = Staff(
            staffId = null,
            staffName = staffName,
            staffRole = staffRole,
            staffPhoto = staffPhoto,
        )

        fun rehydrate(id: Long?, staffName: String, staffRole: String, staffPhoto: String): Staff = Staff(
            staffId = Id.fromNullable(id),
            staffName = staffName,
            staffRole = staffRole,
            staffPhoto = staffPhoto,
        )
    }
}
